### ISSUE TITLE
Bad speculation in JIT_MonExit_Portable

### ISSUE THREAD

**Issue by benaadams** [dotnet/runtime#34800]
There is a high misprediction rate for the common path in JIT_MonExit_Portable

![image](https://user-images.githubusercontent.com/1142958/78956120-3a21b180-7ad9-11ea-9dea-b10ecc10934a.png)

With the common path `action == AwareLock::LeaveHelperAction_None` not being arranged as the common path in asm.

https://github.com/dotnet/runtime/blob/8640eed087d0155fbd76c2ded706b2599e06410a/src/coreclr/src/vm/jithelpers.cpp#L3932-L3936

![image](https://user-images.githubusercontent.com/1142958/78949849-ce354e00-7ac4-11ea-9597-0a724cd3885a.png)

I've tried rearranging this in the C++; however it seems quite resistant to ordering change; perhaps PGO is incorrectly enforcing the order here?

Aside: I assume it needs to be a `cmpxchg` rather than just a volatile write because it may have changed from a lock to a lock+hashcode or perhaps signaling? Would there be any scope to having a lock object type that doesn't allow a hashcode or signaling, so it could just be a volatile write to unlock a non-recursive non-signaled lock? (as `JIT_MonExit_Portable` is quite expensive in the code I'm looking at as 3 highest cost method)

**Comment by jkotas** [comment]
> Would there be any scope to having a lock object type that doesn't allow a hashcode or signaling

Do you mean having a first class lock type (e.g. `System.Threading.Lock`) that is a simple lock without the syncblock baggage? I think that would be useful.

**Comment by benaadams** [comment]
> Do you mean having a first class lock type (e.g. `System.Threading.Lock`) that is a simple lock without the syncblock baggage?

Exactly. Have raised issue https://github.com/dotnet/runtime/issues/34812

**Comment by kouvel** [comment]
The first image seems to indicate that the bad speculation is mostly caused by machine clears and not branch misprediction. I'm not sure what exactly would be causing those, perhaps the `cmpxchg` or false sharing between sync blocks? In .NET 6 too I don't see the ideal code ordering. Was this happening on a benchmark?

> Aside: I assume it needs to be a cmpxchg rather than just a volatile write because it may have changed from a lock to a lock+hashcode or perhaps signaling? Would there be any scope to having a lock object type that doesn't allow a hashcode or signaling, so it could just be a volatile write to unlock a non-recursive non-signaled lock?

The fast paths for exiting `AwareLock` and `LowLevelLock` also currently use CAS to determine if a waiter needs to be signaled. It may be possible to use a volatile write with another full memory barrier instead.
